pseudocode-defns/pseudocode_defns.tex

```diff
@@ -0,0 +1,671 @@
+\end{definition}
+
+\begin{definition}[\texttt{Float}]
+    Generic trait for floating point numbers. 
+\end{definition}
+
+\todo{The Float trait adds \texttt{Div<Self, Output = Self>}}. In general, add the dependencies between traits.
+
+\begin{definition}[\texttt{InfCast}]
+A type \texttt{TO} has trait \texttt{InfCast(TI)} $\iffText$ every cast from a value of type \texttt{TI} to type \texttt{TO} will result in a value of type \texttt{TO} that is at least as big as the value of type \texttt{TI}.
+\end{definition}
+
+\begin{definition}[\texttt{InherentNull}]
+A type \texttt{T} has trait \texttt{InherentNull} $\iffText$ type \texttt{T} can hold some value \texttt{null}.
+
+As of July 16, 2021, only \texttt{f32} and \texttt{f64} have the trait \texttt{InherentNull}.
+\end{definition}
+
+
+\begin{definition}[\texttt{MaxConsecutiveInt}]
+A type $\T$ has trait \texttt{MaxConsecutiveInt} if and only if there is some maximum nonnegative integer \texttt{i} such that all integers from 0 up to \texttt{i} (inclusive) can be expressed as a value of type $\T$; but such that the next integer that can be expressed by $\T$ is not $\texttt{i}+1$.
+\end{definition}
+
+\begin{definition}[\texttt{Metric}]
+A type \texttt{T} has trait \texttt{Metric} $\iffText$ it can represent a metric for quantifying distances between values in a set. The \texttt{Metric} implementation additionally prescribes the type to use for representing distances.
+\end{definition}
+
+\begin{definition}[\texttt{One}]
+    A type $\T$ has trait \texttt{One} $\iffText$ $\T$ has some multiplicative identity element.
+\end{definition}
+
+\begin{definition}[\texttt{OptionNull}]
+A type \texttt{Option<T>} has trait \texttt{OptionNull} $\iffText$ \texttt{null} can be represented as \texttt{Option<T>} \grace{Check this because not sure if it makes sense.}
+\end{definition}
+
+
+\begin{definition}[\texttt{PartialEq}]
+    A type $\T$ has trait \texttt{PartialEq} $\iffText$ $\T$ has some \grace{TODO}
+\end{definition}
+
+\begin{definition}[\texttt{PartialOrd}]\label{def:partial}
+A type \texttt{T} has trait \texttt{PartialOrd} if for all elements $a, b, c$ of type \texttt{T}, the following properties are satisfied:
+
+%A partial ordering over a data domain is a binary relation $\leq$ such that for all $a, b, c$ in this data domain, the following properties are satisfied.
+    \begin{enumerate}
+        \item Reflexivity: $a \leq a$,
+        \item Antisymmetry: if $a \leq b$ and $b \leq a$ then $a=b$,
+        \item Transitivity: if $a \leq b$ and $b \leq c$ then $a \leq c$.
+    \end{enumerate}
+%The partial ordering relation is denoted by \texttt{PartialOrd}.
+\end{definition}
+
+\begin{definition}[\texttt{SaturatingAdd}]
+    A type \texttt{T} has trait \texttt{SaturatingAdd} if it performs addition that saturates at the numeric bounds instead of overflowing.
```

Comment:
Please write more precisely what this means, for example: something like val(a+b)=min{val(a)+val(b),maxval} (using whatever notation is consistent with what you're doing elsewhere).

Reply:
It is written more precisely in definition 4.24. This is a general problem that we were wondering how to resolve: most traits go in pairs with functions. So for example, in the case of saturating add, we have both the trait `SaturatingAdd` and the function `saturating_add`. We decided to write with the most precision in the definition of the function (which is why in this case the trait definition is short), to avoid having the exact same definition. But more generally, how should we go with this?

Reply:
A similar problem occurs for traits that require other traits. E.g., trait `Float` implicitly requires trait `Div<Self, Output = Self>`. But maybe this should not really be part of the definition, so perhaps for each trait we should have a list of the traits that are required.

Reply:
And similarly for functions; e.g., `max` implicitly requires the trait `TotalOrd` in the elements that we apply the function too. This is starting to be a lot of dependencies.

Reply:
I did add a cross-link to the function `saturating_add`. Not sure if there's more to do here. Resolving for now since the initial comment is now fully addressed; feel free to un-resolve.